arrows/ocv/three_frame_differencing.cxx

```diff
@@ -0,0 +1,435 @@
+      cv::threshold( img2_min_minus_img1, img2_min_minus_img1, 0, 1, cv::THRESH_TOZERO );
+      cv::threshold( img1_minus_img2_max, img1_minus_img2_max, 0, 1, cv::THRESH_TOZERO );
+
+      img_diff = cv::max( img2_min_minus_img1, img1_minus_img2_max );
+    }
+  }
+
+  void
+  process_image(cv::Mat &cv_src, cv::Mat &fgmask)
+  {
+    // Images are in temporal order A (oldest), B, C (newest).
+    cv::Mat imgA, imgB, imgC;
+    cv_src.copyTo(imgC);
+    m_frames.push_front(imgC);
+
+    if( m_frames.size() < 2*m_frame_separation )
+    {
+      LOG_TRACE( m_logger, "Haven't collected enough frames yet, so setting "
+                           "foreground mask to all zeros.");
+      fgmask = cv::Mat(cv_src.rows, cv_src.cols, CV_8UC1, cvScalar(0));
+      return;
+    }
+
+    LOG_TRACE( m_logger, "Getting frame from end of queue");
+    imgA = m_frames.back();
+    LOG_TRACE( m_logger, "Getting frame at index frame_separation");
+    imgB = m_frames[m_frame_separation];
+
+    if( m_frames.size() > 2*m_frame_separation )
+    {
+      LOG_TRACE( m_logger, "Removing frame from end of queue");
+      m_frames.pop_back();
+    }
+
+    // unsigned_sum (default)
+    ///  = | | A - C | + | C - B | - | A - B | |
+    cv::Mat AminusC, CminusB, AminusB;
+    image_difference( imgA, imgC, AminusC );
+    image_difference( imgC, imgB, CminusB );
+    image_difference( imgA, imgB, AminusB );
+
+    fgmask = cv::abs( AminusC + CminusB - AminusB );
+
+    if( m_output_to_debug_dir )
+    {
+      std::string fname;
+      cv::Mat img;
+      imgA.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "imgA" + ".tif";
+      cv::imwrite( fname, img );
+      imgB.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "imgB" + ".tif";
+      cv::imwrite( fname, img );
+      imgC.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "imgC" + ".tif";
+      cv::imwrite( fname, img );
+      AminusC.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "AminusC" + ".tif";
+      cv::imwrite( fname, img );
+      CminusB.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "CminusB" + ".tif";
+      cv::imwrite( fname, img );
+      AminusB.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "AminusB" + ".tif";
+      cv::imwrite( fname, img );
+      fgmask.convertTo(img, CV_8UC1);
+      fname = m_debug_dir + "/" + std::to_string(m_debug_counter) + "fgmask" + ".tif";
+      cv::imwrite( fname, img );
+      ++m_debug_counter;
+    }
+
+    //cv::cvtColor(fgmask, fgmask, CV_RGB2GRAY, 1);
+
+    if( fgmask.channels() > 1 )
+    {
+      LOG_TRACE( m_logger, "Converting multichannel foreground mask to single "
+                 "channel");
+      rms_over_channels( fgmask, fgmask, CV_8UC1);
+    }
+
+    if( true )
```

Comment:
Remove ``if( True )`` line if this is to be permanent code or attach to some verbosity parameter.